### PR TITLE
Job Poster Builder step 01 components

### PR DESCRIPTION
**Pull request by Khristophor**
Affects #862, #968.

### Notes:
- Introduces Formik + Yup for easier form state management and validation
- Refactors the Modal component using React portals and context API for more flexibility with the markup and use cases

Still needs localization and tests.
Form element components should be migrated out of `JobDetails.tsx` into their own files for reuse, probably under a `Form` directory or similar.

### CODE REVIEW

**Comment by yonikid15** [comment]
I have made a Form folder in the feature/job-builder/step-0 branch. Moved the Input, Select, etc. into it, while trying to make it a usable as a normal form element or a formik element.

File: resources/assets/js/components/JobDetails/JobDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,734 @@
...
+                      data-c-colour="white"
+                      data-c-font-size="h4"
+                      id="job-details-preview-title"
+                    >
+                      You&apos;re off to a great start!
+                    </h5>
+                  </div>
+                </Modal.Header>
+                <Modal.Body>
+                  <div
```
Comment (tristan-orourke) [inline resources/assets/js/components/JobDetails/JobDetails.tsx]: Extracting the Modal.Body content to a <JobPreviewCard> component might make it clearer what's going on, (and reduce this file size)

Comment (tristan-orourke) [reply resources/assets/js/components/JobDetails/JobDetails.tsx]: And allow it to be reused as the _actual_ job card later

File: resources/assets/js/components/Modal.tsx [hunk 2]
```diff
@@ -1,189 +1,168 @@
-import React, { Component, createRef } from "react";
```
Comment (tristan-orourke) [inline resources/assets/js/components/Modal.tsx]: Meta comment about branch/PR hygene - maybe the changes to the Modal should (ideally) be its own PR

File: resources/assets/js/components/Modal.tsx [hunk 3]
```diff
@@ -1,189 +1,168 @@
...
+    if (modalRef && modalRef.current) {
+      const focusableModalElements = modalRef.current.querySelectorAll(
+        'a[href], button, textarea, input[type="text"], input[type="email"], input[type="radio"], select',
+      );
+      const firstElement = focusableModalElements[0] as HTMLElement;
+      const lastElement = focusableModalElements[
+        focusableModalElements.length - 1
+      ] as HTMLElement;
+
+      if (!e.shiftKey && document.activeElement !== firstElement) {
```
Comment (tristan-orourke) [inline resources/assets/js/components/Modal.tsx]: If I understand correctly, this will always jump between the first and last element. That doesn't work if you add more focusable elements in the modal body.

File: resources/assets/js/components/Modal.tsx [hunk 4]
```diff
@@ -1,189 +1,168 @@
...
+    }
+  }, [children]);
 
-  public handleClose = (event: React.MouseEvent<HTMLButtonElement>): void => {
-    event.preventDefault();
-    this.setState({ visible: false });
-    this.updateBody();
-  };
+  // Adds various key commands to the modal
+  useEffect((): (() => void) => {
```
Comment (tristan-orourke) [inline resources/assets/js/components/Modal.tsx]: Not criticism necessarily, but why does this need to rerun every time children change?

File: resources/assets/js/components/JobDetails/JobDetails.tsx [hunk 5]
```diff
@@ -0,0 +1,734 @@
...
+        "flexHoursOccasionally",
+        "flexHoursSometimes",
+        "flexHoursNever",
+      ],
+      "Please select from the available options",
+    )
+    .required("This field is required."),
+});
+
+const TextInput = ({
```
[the remark this answers is not in the document]
Comment (Khristophor) [reply resources/assets/js/components/JobDetails/JobDetails.tsx]: Formik makes use of some extra props, like 
```js
  field: { name, value, onChange, onBlur },
  form: { errors },
```
we can look at making them optional I suppose, I thought it would be easier to have some separate form components

File: resources/assets/js/components/JobDetails/JobDetails.tsx [hunk 6]
```diff
@@ -0,0 +1,734 @@
...
+    .min(2, "Too Short!")
+    .max(50, "Too Long!")
+    .required("This field is required."),
+  termLength: Yup.number()
+    .min(1, "Too Short!")
+    .max(36, "Too Long!")
+    .required("This field is required."),
+  classification: Yup.mixed()
+    .oneOf(
+      ["AS", "BI", "CO", "CR", "CS", "EC", "EX", "FO", "IS", "PC", "PE", "PM"],
```
[the remark this answers is not in the document]
Comment (Khristophor) [reply resources/assets/js/components/JobDetails/JobDetails.tsx]: I figured we'd be grabbing them from the backend at some point, mostly using this to mock out the frontend at this point.

File: resources/assets/js/components/Modal.tsx (hunk 2 above)
Comment (Khristophor) [reply resources/assets/js/components/Modal.tsx]: It probably should be, although it was driven by the functionality in this story so I thought it was alright in this case.

File: resources/assets/js/components/Modal.tsx (hunk 4 above)
Comment (Khristophor) [reply resources/assets/js/components/Modal.tsx]: On the first run, there aren't any children mounted, so no key commands are wired up correctly. I added the `children` to ensure the event handlers are bound correctly when the content of the modal has settled.

Comment (Khristophor) [reply resources/assets/js/components/Modal.tsx]: If new elements are added/removed from the modal that require focus, the event listeners should run again.

File: resources/assets/js/components/Modal.tsx (hunk 3 above)
Comment (Khristophor) [reply resources/assets/js/components/Modal.tsx]: This just handles the cases where we need to "trap" focus in the modal. Outside of the first or last element in the map, it just tabs through normally.

File: resources/assets/js/components/JobDetails/JobDetails.tsx (hunk 1 above)
Comment (Khristophor) [reply resources/assets/js/components/JobDetails/JobDetails.tsx]: Cool, I can do that.

File: resources/assets/js/components/JobDetails/JobDetails.tsx [hunk 7]
```diff
@@ -0,0 +1,734 @@
...
+                    Next
+                  </button>
+                </div>
+              </Form>
+              <Modal
+                id="job-details-preview"
+                parentElement="#job-details"
+                visible={isModalVisible}
+                onModalConfirm={() => {
+                  console.log("Confirmed");
```
[the remark this answers is not in the document]
Comment (Khristophor) [reply resources/assets/js/components/JobDetails/JobDetails.tsx]: What would you recommend for the interim between components and full Redux functionality?

File: resources/assets/js/components/JobDetails/JobDetails.tsx [hunk 8]
```diff
@@ -0,0 +1,734 @@
...
+                <Field
+                  name="classification"
+                  id="builder02Classification"
+                  label="What is the classification?"
+                  grid="tl(1of2)"
+                  component={SelectInput}
+                  required
+                  nullSelection="Select a classification..."
+                  options={[
+                    { value: "AS", label: "AS - Administrative Services" },
```
[the remark this answers is not in the document]
Comment (Khristophor) [reply resources/assets/js/components/JobDetails/JobDetails.tsx]: Same here, assumed API call for relevant options.

File: resources/assets/js/components/JobDetails/JobDetails.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply resources/assets/js/components/JobDetails/JobDetails.tsx]: Hmm. Yoni did add the field and form props to the Input.tsx component on his branch. But the result is that most props are optional, and a bit confusing. I support having seperate Formik components that wrap the originals.

File: resources/assets/js/components/JobDetails/JobDetails.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply resources/assets/js/components/JobDetails/JobDetails.tsx]: That makes sense. I'm okay with leaving this for this PR, and changing it when we hook it up to the backend.

File: resources/assets/js/components/JobDetails/JobDetails.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply resources/assets/js/components/JobDetails/JobDetails.tsx]: I think JobDetails should accept handler functions as props, to handle things like submit and confirm. In the main app, we'll pass in redux action dispatchers, while the storybook could pass in console.log or anything else.

File: resources/assets/js/components/Modal.tsx (hunk 2 above)
Comment (tristan-orourke) [reply resources/assets/js/components/Modal.tsx]: Yeah, I do this all the time, and I don't expect you to separate it out now. But with this bout of code reviews I've been wondering how we can improve our PR/branch strategies. And I think small feature branches that don't depend on anything except dev, that can be accepted quickly and easily, is the way to go.

File: resources/assets/js/components/JobDetails/JobDetails.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (yonikid15) [reply resources/assets/js/components/JobDetails/JobDetails.tsx]: https://github.com/GCTC-NTGC/TalentCloud/blob/feature/job-builder/step-0/resources/assets/js/components/Forms/Input.tsx 

This is what is looks like currently.

File: resources/assets/js/components/Modal.tsx [hunk 9]
```diff
@@ -1,189 +1,168 @@
...
-        </div>
-      </>
-    );
-  }
+          {children}
+        </modalContext.Provider>
+      </div>
+    </div>,
+    parentElement && parentElement.length > 0
+      ? document.querySelector(parentElement)
```
Comment (tristan-orourke) [inline resources/assets/js/components/Modal.tsx]: It makes me nervous that we pass in a querySelector string into the modal, and hope it works. I'd rather pass in the dom node directly.

File: resources/assets/js/components/Modal.tsx (hunk 3 above)
Comment (Khristophor) [reply resources/assets/js/components/Modal.tsx]: Nope, I'm wrong. Working on refactoring into something that works properly.
